cmd/clusterctl/pkg/log/logger.go

```diff
@@ -0,0 +1,307 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+package log implements a clusterctl friendly logr.Logger derived from
+https://github.com/kubernetes/klog/blob/master/klogr/klogr.go.
+
+The logger is designed to work in two scenarios:
+
+1. clusterctl CLI, with logs printed to stdout with a formatting that is easy to read for users
+   but also simple to parse for identifying specific values.
+
+2. clusterctl library, with log entries sent to a channel for allowing different types of consumption.
+
+Note: the clusterctl library also support usage of other loggers as long as they conform to the github.com/go-logr/logr.Logger interface.
+
+Following logging conventions are used in clusterctl:
+
+Message:
+
+All messages should start with a capital letter.
+
+Log level:
+
+Use Level 0 (the default, if you don't specify a level) for the most important user feedback only, e.g.
+- reporting command progress for long running actions
+- reporting command results when required
+
+Use logging Levels 1 providing more info about the command's internal workflow.
+
+Use logging Levels 5 for for providing all the information required for debug purposes/problem investigation.
+
+Logging WithValues:
+
+Logging WithValues should be preferred to embedding values into log messages because it allows
+machine readability.
+
+Variables name should start with a capital letter.
+
+Logging WithNames:
+
+Logging WithNames should be used carefully.
+Please consider that practices like prefixing the logs with something indicating which part of code
+is generating the log entry might be useful for developers, but it can create confusion for
+the end users because it increases the verbosity without providing information the user can understand/take benefit from.
+
+Logging errors:
+
+A proper error management should always be preferred to the usage of log.Error.
+
+*/
+
+package log
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// LogEntry defines the information that can be used for composing a log line.
+type LogEntry struct {
+	// Prefix of the log line, composed of the hierarchy of log.WithName values.
+	Prefix string
+
+	// Level of the LogEntry.
+	Level int
+
+	// Values of the log line, composed of the concatenation of log.WithValues and KeyValue pairs passed to log.Info.
+	Values []interface{}
+}
+
+// LogChannel define a channel where LogEntry can be sent.
+type LogChannel chan LogEntry
+
+// NewOptions carries the options supported by NewLogger
+type NewOptions struct {
+	ch        LogChannel
+	threshold int
+}
```

Comment:
Please remove this and replace it with...

Reply:
done